src/mappings/permissionedExchange.ts

```diff
@@ -0,0 +1,139 @@
+import { BigNumber } from 'ethers';
+
+const { ACTIVE, INACTIVE } = OrderStatus;
+
+function calculateTradeAmount(
+    totalTradeAmount: bigint,
+    event: AcalaEvmEvent<TradeEvent['args']>
+): bigint {
+    const { tokenGive, tokenGet, amountGive, amountGet } = event.args;
+    const getIsKSQT = tokenGet === 'kSQT';
+    const giveIsKSQT = tokenGive === 'kSQT';
+
+    if(giveIsKSQT && getIsKSQT) return totalTradeAmount;
+
+    const tradeAmountBN = BigNumber.from(totalTradeAmount);
+
+    if(giveIsKSQT) return tradeAmountBN.add(amountGive).toBigInt();
+    if(getIsKSQT) return tradeAmountBN.sub(amountGet).toBigInt();
+
+    return totalTradeAmount;
+}
+
+async function createTrade(
+    orderId: BigNumber,
+    sender: string,
+    event: AcalaEvmEvent<TradeEvent['args']>
+): Promise<void> {
+    const { tokenGive, tokenGet, amountGive, amountGet } = event.args;
+
+    const trade = Trade.create({
+        id: `${orderId.toString()}:${event.transactionHash}`,
+        tokenGive,
+        tokenGet,
+        amountGive: amountGet.toBigInt(),
+        amountGet: amountGive.toBigInt(),
+        senderId: sender
+    });
+
+    await trade.save();   
+}
+
+export async function handleExchangeOrderSent(
+    event: AcalaEvmEvent<ExchangeOrderSentEvent['args']>
+  ): Promise<void> {
+    logger.info('handleRewardsClaimed');
+    assert(event.args, 'No event args');
+
+    const { orderId, sender, tokenGive, tokenGet, amountGive, amountGet, expireDate } = event.args;
+
+    const order = Order.create({
+        id: orderId.toString(),
+        sender,
+        tokenGive,
+        tokenGet,
+        amountGive: amountGive.toBigInt(),
+        amountGet: amountGet.toBigInt(),
+        expireDate: new Date(expireDate.toNumber()),
+        amountGiveLeft: amountGive.toBigInt(),
+        status: ACTIVE
+    });
+
+    await order.save();
+}
+
+export async function handleTrade(
+    event: AcalaEvmEvent<TradeEvent['args']>
+){
+    logger.info('handleTrade');
+    assert(event.args, 'No event args');
+
+    const { orderId } = event.args; 
+    const handlerInfo = getUpsertAt('handleOrderSettled', event);
+
+    const permissionedExchange = PermissionedExchange__factory.connect(
+        EXCHANGE_DIST_ADDRESS,
+        new FrontierEthProvider()
+    );
+    
+    //FIXME: I may not need to get amountGiveLeft every time I have a trade event.
+    const { amountGiveLeft } = await permissionedExchange.orders(orderId);    
+    const sender = event.from;
+
+    await createTrade(orderId, sender, event);
+
+    //-- Order Entity handling 
+
+    const order = await Order.get(orderId.toString());
+    assert(order, `Expect order with id ${orderId.toString()} to exist`);
+
+    order.amountGiveLeft = amountGiveLeft.toBigInt();
+    await order.save();
+
+    //-- Trader Entity handling 
+
+    let trader = await Trader.get(sender);
+    const totalTradeAmount = calculateTradeAmount(trader ? trader.totalTradeAmount : BigInt(0), event);
```

Comment:
`totalTradeAmount` looks special for `ksqt`, in that case, just check whether the token give is `ksqt` than add the `amountGive`

Reply:
Updated to I need to account for if both tokenGive and tokenGet are kSQT?

Reply:
We control the order so the case will not happen